Python/chapter03/p01_three_in_one/miguelHx.py

```diff
@@ -0,0 +1,445 @@
+"""Python version 3.7.0
+3.1 - Three in one
+Describe how you could use a single array to implement three stacks
+"""
+
+import copy
+import unittest
+
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+from typing import List, Optional, Iterator
+
+T = TypeVar('T')
+
+@dataclass
+class StackNode(Generic[T]):
+    data: T
+    next: 'Optional[StackNode[T]]'
+
+
+class MyStack:
+    """Stack data structure implementation.
+
+    Uses LIFO (last-in first-out) ordering.
+    The most recent item added to the stack is
+    the first removed.  Traversal is top to bottom.
+    """
+
+    def __init__(self):
+        self.top: Optional[StackNode[T]] = None # top is a pointer to StackNode object
+        self.size: int = 0
+        self.index: int = -1 # tracking current node for iterator use
+        self.current_node: Optional[StackNode[T]] = self.top
+        return
+    
+    def pop(self) -> T:
+        """
+        Removes the top item from the stack
+
+        Raises:
+            IndexError: raised when pop is attempted on empty stack
+
+        Returns:
+            int: The data at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        self.current_node = self.top
+        self.size -= 1
+        return item
+    
+    def push(self, item: int) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item, None)
+        t.next = self.top
+        self.top = t
+        self.current_node = self.top
+        self.size += 1
+
+    def peek(self) -> int:
+        """
+        Returns data at the top of the stack
+
+        Raises:
+            IndexError: [description]
+
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+    
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return self
+    
+    def __next__(self) -> T:
+        self.index += 1
+        if self.index == self.size or self.current_node is None:
+            self.index = -1
+            self.current_node = self.top
+            raise StopIteration
+        n: T = self.current_node.data
+        self.current_node = self.current_node.next
+        return n
+    
+    def __bool__(self) -> bool:
+        """
+        Returns True if and only if the
+        stack is empty, otherwise False.
```

Comment:
idiomatically in Python, `True` is returned when the container is _not_ empty. From https://docs.python.org/3/reference/datamodel.html#object.__bool__ : 

> Called to implement truth value testing and the built-in operation bool(); should return False or True. When this method is not defined, __len__() is called, if it is defined, and the object is considered true if its result is nonzero. If a class defines neither __len__() nor __bool__(), all its instances are considered true.

Reply:
Good catch.  I somehow forgot to add proper test cases for __bool__.

Resolved in e07edab577b0c84bee942fd9e7d8975adb9e9376